velox/functions/prestosql/tests/ArrayShuffleTest.cpp

```diff
@@ -0,0 +1,316 @@
+
+      ASSERT_TRUE(std::is_permutation(
+          actualArray.begin(), actualArray.end(), expectedArray.begin()))
+          << "Actual array " << printArray(actualArray) << " at " << i
+          << " must produce a permutation of expected array "
+          << printArray(expectedArray);
+
+      distinctValueSet.insert(actualArray);
+    }
+
+    // Shuffled arrays should be X% different in total.
+    const int32_t kThreshold =
+        (int32_t)(actualVector->size() * kUniquenessRate);
+    auto numDistinctValues = distinctValueSet.size();
+    ASSERT_TRUE(numDistinctValues >= kThreshold)
+        << "Shuffle " << input->encoding()
+        << " array must yield >= " << kThreshold
+        << " distinct values, but only got " << numDistinctValues;
+  }
+};
+} // namespace
+
+TEST_F(ArrayShuffleTest, normalArrays) {
+  auto input = makeNullableArrayVector<int64_t>(
+      {{},
+       {std::nullopt},
+       {std::nullopt, std::nullopt},
+       {-1, 0, 1},
+       {std::nullopt, 0, 0},
+       {std::numeric_limits<int64_t>::max(),
+        std::numeric_limits<int64_t>::min(),
+        0,
+        1,
+        2,
+        3}});
+  testShuffle<int64_t, true>(input);
+}
+
+TEST_F(ArrayShuffleTest, nestedArrays) {
+  using innerArrayType = std::vector<std::optional<int64_t>>;
+  using outerArrayType =
+      std::vector<std::optional<std::vector<std::optional<int64_t>>>>;
+  innerArrayType a{1, 2, 3, 4};
+  innerArrayType b{5, 6};
+  innerArrayType c{6, 7, 8};
+  outerArrayType row1{{a}, {b}};
+  outerArrayType row2{std::nullopt, std::nullopt, {a}, {b}, {c}};
+  outerArrayType row3{{{}}};
+  outerArrayType row4{{{std::nullopt}}};
+  auto input =
+      makeNullableNestedArrayVector<int64_t>({{row1}, {row2}, {row3}, {row4}});
+
+  testShuffle<Array<int64_t>, true>(input);
+}
+
+TEST_F(ArrayShuffleTest, sortAndShuffle) {
+  auto input = makeNullableArrayVector<int64_t>(
+      {{-1, 0, std::nullopt, 1, std::nullopt},
+       {4, 1, 5, 3, 2},
+       {std::numeric_limits<int64_t>::max(),
+        std::numeric_limits<int64_t>::min(),
+        4,
+        1,
+        3,
+        2,
+        std::nullopt}});
+  auto inputVector = makeRowVector({input});
+  auto result1 = evaluate<ArrayVector>("array_sort(C0)", inputVector);
+  auto result2 = evaluate<ArrayVector>("array_sort(shuffle(C0))", inputVector);
+
+  assertEqualVectors(result1, result2);
+}
+
+TEST_F(ArrayShuffleTest, constantEncoding) {
+  vector_size_t size = 100;
+  // Considering the cases of empty array, nullable array, duplicate-element
+  // array, and distinct-value array.
+  auto valueVector = makeNullableArrayVector<int64_t>(
+      {{}, {std::nullopt, 0}, {5, 5}, {1, 2, 3}});
+
+  for (auto i = 0; i < valueVector->size(); i++) {
+    auto input = BaseVector::wrapInConstant(size, i, valueVector);
+    testShuffle<int64_t, true>(input);
+  }
+}
+
+TEST_F(ArrayShuffleTest, dictEncoding) {
+  // Considering the case of repeated elements: {1,2,3} x 3, {4,5} x 2.
```

Comment:
This use case doesn't appear to be tested. Would you double check?

Reply:
What do you mean "doesn't appear to be tested"? 
Right now, I'm using std::is_permutation() to test if the shuffled output is a permutation of the origin input in test_shuffle() for the **correctness purposes**.
Regarding the **randomness test**, you will need to look at the test case: `dictEncodingRandomness`. It will generate a dictionary with size=200 base data and size=100 indices. The dictionary values will all be identical: {0,1,2,...,9}. And it will test if the shuffled output contains at least 7 distinct permutations (e.g. {1,2,3},{1,2,3},... => {1,3,2}, {3,2,1}). Putting in another word, if it shuffles only once for those repeated dict elements, it will yield one permutation (e.g. {1,2,3} => {1,3,2}). Does this new test case cover your question?